File: lecture_summarizer/render/subtitles.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, List


def _format_timestamp(seconds: float, *, kind: str) -> str:
    total_ms = max(int(round(seconds * 1000)), 0)
    hours, remainder = divmod(total_ms, 3_600_000)
    minutes, remainder = divmod(remainder, 60_000)
    sec, ms = divmod(remainder, 1000)
    if kind == "srt":
        return f"{hours:02d}:{minutes:02d}:{sec:02d},{ms:03d}"
    return f"{hours:02d}:{minutes:02d}:{sec:02d}.{ms:03d}"


def _split_text(text: str, parts: int) -> List[str]:
    cleaned = text.strip()
    if not cleaned:
        return [""] * max(parts, 1)
    paragraphs = [p.strip() for p in cleaned.split("\n\n") if p.strip()]
    if not paragraphs:
        paragraphs = [cleaned]
    if parts <= 1:
        return ["\n".join(paragraphs)]
    segments: List[str] = [""] * parts
    for idx, paragraph in enumerate(paragraphs):
        target = idx % parts
        if segments[target]:
            segments[target] += "\n\n"
        segments[target] += paragraph
    return segments


@dataclass
class SubtitleExporter:
    """Generate simple SRT and VTT exports from aggregated transcript text."""

    def write(
        self,
        fmt: str,
        *,
        base: Path,
        name: str,
        text: str,
        chunks: Iterable[dict] | None,
    ) -> Path | None:
        fmt_lower = fmt.lower()
        if fmt_lower not in {"srt", "vtt"}:
            return None
        base.mkdir(parents=True, exist_ok=True)
        target = base / f"{name}.{fmt_lower}"
        chunk_list = list(chunks or [])
        if not chunk_list:
            chunk_list = [{"start_seconds": 0.0, "end_seconds": 0.0}]
        segments = _split_text(text, len(chunk_list))
        lines: List[str] = []
        if fmt_lower == "vtt":
            lines.append("WEBVTT\n")
        for idx, chunk in enumerate(chunk_list):
            start = float(chunk.get("start_seconds", idx * 5.0))
            end = float(chunk.get("end_seconds", start + 5.0))
            content = segments[idx] if idx < len(segments) else ""
            if fmt_lower == "srt":
                lines.append(str(idx + 1))
                lines.append(f"{_format_timestamp(start, kind='srt')} --> {_format_timestamp(end, kind='srt')}")
                lines.append(content or "[No content]")
                lines.append("")
            else:
                lines.append(f"{_format_timestamp(start, kind='vtt')} --> {_format_timestamp(end, kind='vtt')}")
                lines.append(content or "[No content]")
                lines.append("")
        target.write_text("\n".join(lines).strip() + "\n", encoding="utf-8")
        return target

File: src/render/subtitles.rs
use anyhow::Result;
use serde_json::Value;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Default, Clone)]
pub struct SubtitleExporter;

impl SubtitleExporter {
    pub fn write(
        &self,
        fmt: &str,
        base: &Path,
        name: &str,
        text: &str,
        chunks: &[Value],
    ) -> Result<Option<PathBuf>> {
        let fmt = fmt.trim().to_lowercase();
        if fmt != "srt" && fmt != "vtt" {
            return Ok(None);
        }
        fs::create_dir_all(base)?;
        let target = base.join(format!("{name}.{fmt}"));
        let segments = split_text(text, chunks.len());
        let mut lines = Vec::new();
        if fmt == "vtt" {
            lines.push("WEBVTT".to_string());
            lines.push(String::new());
        }
        for (idx, chunk) in chunks.iter().enumerate() {
            let segment = segments.get(idx).cloned().unwrap_or_default();
            let start = chunk
                .get("start_seconds")
                .and_then(Value::as_f64)
                .unwrap_or((idx * 5) as f64);
            let end = chunk
                .get("end_seconds")
                .and_then(Value::as_f64)
                .unwrap_or(start + 5.0);
            if fmt == "srt" {
                lines.push((idx + 1).to_string());
                lines.push(format!(
                    "{} --> {}",
                    format_timestamp(start, Format::Srt),
                    format_timestamp(end, Format::Srt)
                ));
                lines.push(if segment.is_empty() {
                    "[No content]".to_string()
                } else {
                    segment
                });
                lines.push(String::new());
            } else {
                lines.push(format!(
                    "{} --> {}",
                    format_timestamp(start, Format::Vtt),
                    format_timestamp(end, Format::Vtt)
                ));
                lines.push(if segment.is_empty() {
                    "[No content]".to_string()
                } else {
                    segment
                });
                lines.push(String::new());
            }
        }
        if chunks.is_empty() {
            let start = 0.0;
            let end = 5.0;
            if fmt == "srt" {
                lines.push("1".into());
                lines.push(format!(
                    "{} --> {}",
                    format_timestamp(start, Format::Srt),
                    format_timestamp(end, Format::Srt)
                ));
                lines.push(text.trim().to_string());
            } else {
                lines.push(format!(
                    "{} --> {}",
                    format_timestamp(start, Format::Vtt),
                    format_timestamp(end, Format::Vtt)
                ));
                lines.push(text.trim().to_string());
            }
        }
        std::fs::write(&target, lines.join("\n"))?;
        Ok(Some(target))
    }
}

fn split_text(text: &str, parts: usize) -> Vec<String> {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return vec![String::new(); parts.max(1)];
    }
    let paragraphs = trimmed
        .split("\n\n")
        .map(|p| p.trim())
        .filter(|p| !p.is_empty())
        .collect::<Vec<_>>();
    if parts <= 1 || paragraphs.is_empty() {
        return vec![paragraphs.join("\n\n")];
    }
    let mut segments = vec![String::new(); parts];
    for (idx, para) in paragraphs.iter().enumerate() {
        let slot = idx % parts;
        if !segments[slot].is_empty() {
            segments[slot].push_str("\n\n");
        }
        segments[slot].push_str(para);
    }
    segments
}

#[derive(Copy, Clone)]
enum Format {
    Srt,
    Vtt,
}

fn format_timestamp(seconds: f64, fmt: Format) -> String {
    let total_ms = (seconds.max(0.0) * 1000.0).round() as i64;
    let hours = total_ms / 3_600_000;
    let minutes = (total_ms % 3_600_000) / 60_000;
    let secs = (total_ms % 60_000) / 1000;
    let millis = total_ms % 1000;
    match fmt {
        Format::Srt => format!("{hours:02}:{minutes:02}:{secs:02},{millis:03}"),
        Format::Vtt => format!("{hours:02}:{minutes:02}:{secs:02}.{millis:03}"),
    }
}
